### PR TITLE
new logging

### PR DESCRIPTION
**Pull request by graydon**
This is a sketch of migration to a new logging system: spdlog. It uses a fmt-based interface. It can be enabled with --enable-spdlog. For the time being it does not yet remove easylogging++, which remains the default. I'll post a followup to remove the latter, later.

### CODE REVIEW

**Comment by graydon** [comment]
comparison of full (opt) builds:

```
quill:
real    6m4.354s
user    117m6.541s
sys     3m8.383s
-rwxrwxr-x 1 graydon graydon 143142528 Dec 10 18:20 src/stellar-core
-rwxrwxr-x 1 graydon graydon  23793184 Dec 10 18:21 src/stellar-core (stripped)


spdlog:
real    5m35.381s
user    98m26.085s
sys     2m38.941s
-rwxrwxr-x 1 graydon graydon 120296624 Dec 10 18:37 src/stellar-core
-rwxrwxr-x 1 graydon graydon  19508672 Dec 10 18:38 src/stellar-core
```

**Comment by graydon** [comment]
comparison of runtime performance:

```
 Performance counter stats for 'src/stellar-core-easylogging --ll info test [flood]' (3 runs):

           4738.94 msec task-clock                #    0.551 CPUs utilized            ( +-  0.50% )
               770      context-switches          #    0.162 K/sec                    ( +-  2.02% )
                10      cpu-migrations            #    0.002 K/sec                    ( +- 11.55% )
             51183      page-faults               #    0.011 M/sec                    ( +-  0.46% )
       13236053148      cycles                    #    2.793 GHz                      ( +-  0.21% )  (60.01%)
        7365920838      stalled-cycles-frontend   #   55.65% frontend cycles idle     ( +-  0.18% )  (59.86%)
       14850807138      instructions              #    1.12  insn per cycle         
                                                  #    0.50  stalled cycles per insn  ( +-  0.21% )  (60.00%)
        2010600000      branches                  #  424.272 M/sec                    ( +-  0.31% )  (60.16%)
          53937888      branch-misses             #    2.68% of all branches          ( +-  0.19% )  (59.97%)

            8.6069 +- 0.0501 seconds time elapsed  ( +-  0.58% )


 Performance counter stats for 'src/stellar-core-easylogging --ll trace test [flood]' (3 runs):

           8213.43 msec task-clock                #    0.692 CPUs utilized            ( +-  2.22% )
               793      context-switches          #    0.097 K/sec                    ( +-  4.36% )
                 5      cpu-migrations            #    0.001 K/sec                    ( +- 31.13% )
             63080      page-faults               #    0.008 M/sec                    ( +-  0.56% )
       23974128502      cycles                    #    2.919 GHz                      ( +-  0.72% )  (60.00%)
       14717286633      stalled-cycles-frontend   #   61.39% frontend cycles idle     ( +-  0.80% )  (60.07%)
       22211947500      instructions              #    0.93  insn per cycle         
                                                  #    0.66  stalled cycles per insn  ( +-  0.23% )  (60.06%)
        3337141619      branches                  #  406.303 M/sec                    ( +-  0.44% )  (59.92%)
          95470031      branch-misses             #    2.86% of all branches          ( +-  0.59% )  (59.95%)

            11.877 +- 0.198 seconds time elapsed  ( +-  1.66% )


 Performance counter stats for 'src/stellar-core-quill --ll info test [flood]' (3 runs):

           5228.41 msec task-clock                #    0.627 CPUs utilized            ( +-  1.07% )
            131289      context-switches          #    0.025 M/sec                    ( +-  2.37% )
                58      cpu-migrations            #    0.011 K/sec                    ( +- 21.56% )
             63986      page-faults               #    0.012 M/sec                    ( +-  2.07% )
       13779875739      cycles                    #    2.636 GHz                      ( +-  1.91% )  (60.93%)
        7649167200      stalled-cycles-frontend   #   55.51% frontend cycles idle     ( +-  2.48% )  (61.35%)
       15656028696      instructions              #    1.14  insn per cycle         
                                                  #    0.49  stalled cycles per insn  ( +-  0.73% )  (59.21%)
        2211589730      branches                  #  422.995 M/sec                    ( +-  0.62% )  (58.78%)
          54878566      branch-misses             #    2.48% of all branches          ( +-  0.57% )  (59.74%)

            8.3444 +- 0.0807 seconds time elapsed  ( +-  0.97% )


 Performance counter stats for 'src/stellar-core-quill --ll trace test [flood]' (3 runs):

           6015.43 msec task-clock                #    0.693 CPUs utilized            ( +-  0.87% )
            130846      context-switches          #    0.022 M/sec                    ( +-  0.38% )
                63      cpu-migrations            #    0.010 K/sec                    ( +-  6.91% )
             68883      page-faults               #    0.011 M/sec                    ( +-  1.39% )
       16191431998      cycles                    #    2.692 GHz                      ( +-  0.32% )  (59.94%)
        8948372899      stalled-cycles-frontend   #   55.27% frontend cycles idle     ( +-  0.61% )  (60.50%)
       18468143381      instructions              #    1.14  insn per cycle         
                                                  #    0.48  stalled cycles per insn  ( +-  0.28% )  (59.95%)
        2651516608      branches                  #  440.786 M/sec                    ( +-  0.44% )  (59.42%)
          62353109      branch-misses             #    2.35% of all branches          ( +-  0.10% )  (60.20%)

            8.6745 +- 0.0216 seconds time elapsed  ( +-  0.25% )


 Performance counter stats for 'src/stellar-core-spdlog --ll info test [flood]' (3 runs):

           4353.73 msec task-clock                #    0.529 CPUs utilized            ( +-  2.06% )
               759      context-switches          #    0.174 K/sec                    ( +-  0.34% )
                 6      cpu-migrations            #    0.001 K/sec                    ( +- 21.21% )
             44439      page-faults               #    0.010 M/sec                    ( +-  9.90% )
       12299892666      cycles                    #    2.825 GHz                      ( +-  0.08% )  (59.76%)
        6758492997      stalled-cycles-frontend   #   54.95% frontend cycles idle     ( +-  0.03% )  (59.87%)
       14215487786      instructions              #    1.16  insn per cycle         
                                                  #    0.48  stalled cycles per insn  ( +-  0.20% )  (60.09%)
        1871349752      branches                  #  429.827 M/sec                    ( +-  0.10% )  (60.21%)
          49035179      branch-misses             #    2.62% of all branches          ( +-  0.07% )  (60.08%)

            8.2287 +- 0.0893 seconds time elapsed  ( +-  1.08% )


 Performance counter stats for 'src/stellar-core-spdlog --ll trace test [flood]' (3 runs):

           5722.83 msec task-clock                #    0.603 CPUs utilized            ( +-  1.51% )
               752      context-switches          #    0.131 K/sec                    ( +-  0.39% )
                10      cpu-migrations            #    0.002 K/sec                    ( +- 26.46% )
             70070      page-faults               #    0.012 M/sec                    ( +-  4.03% )
       17097807994      cycles                    #    2.988 GHz                      ( +-  0.96% )  (59.92%)
        9890556258      stalled-cycles-frontend   #   57.85% frontend cycles idle     ( +-  0.48% )  (59.81%)
       18149677490      instructions              #    1.06  insn per cycle         
                                                  #    0.54  stalled cycles per insn  ( +-  2.39% )  (60.01%)
        2488373624      branches                  #  434.815 M/sec                    ( +-  1.86% )  (60.22%)
          67135727      branch-misses             #    2.70% of all branches          ( +-  1.62% )  (60.03%)

            9.4853 +- 0.0825 seconds time elapsed  ( +-  0.87% )

```

**Comment by graydon** [comment]
Summary of the above:

  - Both spdlog and quill are improvements in perf over easylogging
  - Quill takes significantly longer to compile and produces significantly larger binaries
  - Spdlog is very slightly faster than quill when lightly loaded (at coarse info-level logging)
  - Quill degrades only very slightly when switching to higher load (finer grained trace-level)
  - Spdlog degrades more noticably going from info to trace, but is still much faster than easylogging
  - Quill achieves its speed by consuming more on-core time and more context switches overall (i.e. "because it's async and uses a background thread")

Given these measurements I'm leaning towards spdlog, but will work on finishing the conversion / checking for other quirks in the implementation (especially around file handling) before a definitive choice.

**Comment by graydon** [comment]
(updated size numbers based on inclusion of FMT_STRING which is used for compile-time fmt checking)

**Comment by MonsieurNicolas** [comment]
From what you're saying, I am also leaning towards spdlog that seems to strike the right balance, and if we use own own wrapper everywhere, we can easuily switch to something else later (`fmt::format` is becoming the standard).

Now, as for the way the end results looks like:
```
CLOG(TRACE, "Bucket") << h << " referenced by LCL";
```
gets translated into
```
LOG_TRACE(GET_LOG("Bucket"), "{} referenced by LCL", h);
```
why not something like?
```
CLOG(TRACE, "Bucket", "{} referenced by LCL", h);
```
You can surely write a `CLOG` macro that would do the conversion `LOG_X(GET_LOG(Y), ...)` and avoid all this boiler plate stuff.

Also, your script(?) produces some weird stuff (in this example, there was no `fmt::format` in the original log, in the few places where we do, we should also fix them up):
```
LOG_TRACE(GET_LOG("Bucket"),
            "BucketMergeMap forgetting mappings for merge {} <-> output={}",
            fmt::format("{}", mergeKeyProducingOutput),
            hexAbbrev(outputBeingDropped));
```
We should never have to nest `fmt::format` calls.

**Comment by graydon** [comment]
I could do that macro -- but my hope was to go through with a later PR and modify most of the GET_LOG calls to happen in method initializers, and cache the logger objects inside local objects. That's one of the wins of this whole approach.

The nested format calls to mergeKeys are -- I think -- a transitionary artifact when I was trying to work out how to make operator<< interoperate. I think they may be necessary for quill, but I'm leaning away from bothering with quill anyways. I will have a look at them.

**Comment by graydon** [comment]
Ah right: the nested format calls are because quill wants to make a copy of the _unformatted_ object to send to its formatting backend and your only options are to pre-format the object yourself or mark the object with QUILL_COPY_LOGGABLE. I will do the latter -- I just didn't see the page explaining how to do it at first.

**Comment by MonsieurNicolas** [comment]
> I could do that macro -- but my hope was to go through with a later PR and modify most of the GET_LOG calls to happen in method initializers, and cache the logger objects inside local objects. That's one of the wins of this whole approach.

I think we touched on this before: I don't think I'd want to have this kind of thing stored on a per class basis/pass around loggers to free functions. This is one of the few times when a thread local storage (if you even need that) global is acceptable and actually better.
Also, we should avoid doing another giant PR that touches most files, so this PR should be the one doing that.

Related: in code, do we even need the partition name to be a string? Maybe it should just be something like `CLOG(INFO, BUCKET, "format string", params...);` that way we can use strings, variable mangling, or an enum (index in an array) in the macro that actually "does the right thing".

On `QUILL_COPY_LOGGABLE`: if we use `spdlog`, we probably don't need that change either, right?

**Comment by graydon** [comment]
Correct, spdlog doesn't require any of this. Maybe I'll just purge the quill stuff and move ahead with spdlog.

**Comment by graydon** [comment]
Pushed variant that removes quill, refactors init/deinit code, adds submodule, supports rotation, adds optional but not default support for color, and (as far as I can tell) is indistinguishable from easylogging while running. Still treats spdlog as optional. Will post later PR that commits fully & removes easylogging.

**Comment by graydon** [comment]
(I believe this is ready to merge, if tests pass)

File: src/bucket/BucketList.cpp [hunk 1]
```diff
@@ -111,8 +111,9 @@ BucketLevel::commit()
     if (mNextCurr.isLive())
     {
         setCurr(mNextCurr.resolve());
-        // CLOG(DEBUG, "Bucket") << "level " << mLevel << " set mCurr to "
-        //            << mCurr->getEntries().size() << " elements";
+        // CLOG_DEBUG(Bucket, "level {}{}{} elements", mLevel, " set
```
Comment (MonsieurNicolas) [inline src/bucket/BucketList.cpp:114]: looks like your script didn't work here. I'd just remove it or convert it to a TRACE level log

File: src/bucket/BucketList.cpp [hunk 2]
```diff
@@ -179,10 +180,12 @@ BucketLevel::snap()
 {
     mSnap = mCurr;
     mCurr = std::make_shared<Bucket>();
-    // CLOG(DEBUG, "Bucket") << "level " << mLevel << " set mSnap to "
-    //            << mSnap->getEntries().size() << " elements";
-    // CLOG(DEBUG, "Bucket") << "level " << mLevel << " reset mCurr to "
-    //            << mCurr->getEntries().size() << " elements";
+    // CLOG_DEBUG(Bucket, "level {}{}{} elements", mLevel, " set mSnap
```
Comment (MonsieurNicolas) [inline src/bucket/BucketList.cpp:183]: same thing: commented out logging statements are not useful. remove or TRACE

File: src/bucket/BucketList.cpp [hunk 3]
```diff
@@ -522,11 +525,10 @@ BucketList::addBatch(Application& app, uint32_t currLedger,
         shadows.pop_back();
 
         /*
-        CLOG(DEBUG, "Bucket") << "curr=" << currLedger
-                   << ", half(i-1)=" << levelHalf(i-1)
-                   << ", size(i-1)=" << levelSize(i-1)
-                   << ", mask(curr,half)=" << mask(currLedger, levelHalf(i-1))
-                   << ", mask(curr,size)=" << mask(currLedger, levelSize(i-1));
+        CLOG_DEBUG(Bucket, "curr={}, half(i-1)={}, size(i-1)={},
+        mask(curr,half)={}, mask(curr,size)={}", currLedger, levelHalf(i-1),
```
Comment (MonsieurNicolas) [inline src/bucket/BucketList.cpp:529]: same

File: src/bucket/BucketList.cpp [hunk 4]
```diff
@@ -557,12 +559,13 @@ BucketList::addBatch(Application& app, uint32_t currLedger,
 
             auto snap = mLevels[i - 1].snap();
 
-            // CLOG(DEBUG, "Bucket") << "Ledger " << currLedger
-            //           << " causing commit on level " << i
-            //           << " and prepare of "
-            //           << snap->getEntries().size()
-            //           << " element snap from level " << i-1
-            //           << " to level " << i;
+            // CLOG_DEBUG(Bucket, "Ledger {} causing commit on level
```
Comment (MonsieurNicolas) [inline src/bucket/BucketList.cpp:562]: more

File: src/bucket/test/BucketListTests.cpp [hunk 5]
```diff
@@ -148,17 +148,17 @@ TEST_CASE("bucket list", "[bucket][bucketlist]")
...
                 for (uint32_t j = 0; j < BucketList::kNumLevels; ++j)
                 {
                     auto const& lev = bl.getLevel(j);
                     auto currSz = countEntries(lev.getCurr());
                     auto snapSz = countEntries(lev.getSnap());
-                    // CLOG(DEBUG, "Bucket") << "level " << j
-                    //            << " curr=" << currSz
-                    //            << " snap=" << snapSz;
+                    // CLOG_DEBUG(Bucket, "level {} curr={} snap={}",
+                    // j
```
Comment (MonsieurNicolas) [inline src/bucket/test/BucketListTests.cpp:159]: remove or TRACE

File: src/bucket/test/BucketListTests.cpp [hunk 6]
```diff
@@ -645,38 +643,30 @@ TEST_CASE("BucketList number dump", "[bucket][bucketlist][count][!hide]")
 {
     for (uint32_t level = 0; level < BucketList::kNumLevels; ++level)
     {
-        CLOG(INFO, "Bucket")
-            << "levelSize(" << level
-            << ") = " << formatU32(BucketList::levelSize(level))
-            << " (formally)";
+        CLOG_INFO(Bucket, "levelSize({}) = {} (formally)", level,
```
Comment (MonsieurNicolas) [inline src/bucket/test/BucketListTests.cpp:646]: looks like `formatU32` is now obsolete, can just be done with the format string?

File: src/util/Logging.cpp [hunk 7]
```diff
@@ -111,18 +258,23 @@ Logging::getStringFromLL(el::Level level)
...
 Logging::logTrace(std::string const& partition)
 {
+    std::lock_guard<std::recursive_mutex> guard(mLogMutex);
     return mLogLevel <= el::Level::TRACE;
 }
 
 void
 Logging::rotate()
 {
+    std::lock_guard<std::recursive_mutex> guard(mLogMutex);
```
Comment (MonsieurNicolas) [inline src/util/Logging.cpp:275]: Does it mean that when rotating logs it forgets the current log level? I see that `setLogLevel` performs a bunch of stuff that we're not redoing in `init`

File: src/bucket/BucketList.cpp (hunk 1 above)
Comment (MonsieurNicolas) [reply src/bucket/BucketList.cpp:114]: you can look for `// CLOG` and `// LOG` to find all those places

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
question on log rotation, otherwise minor suggestions. This looks great

File: src/bucket/BucketList.cpp (hunk 1 above)
Comment (graydon) [reply src/bucket/BucketList.cpp:114]: fixed.

File: src/bucket/BucketList.cpp (hunk 2 above)
Comment (graydon) [reply src/bucket/BucketList.cpp:183]: fixed.

File: src/bucket/BucketList.cpp (hunk 3 above)
Comment (graydon) [reply src/bucket/BucketList.cpp:529]: fixed.

File: src/bucket/BucketList.cpp (hunk 4 above)
Comment (graydon) [reply src/bucket/BucketList.cpp:562]: fixed

File: src/bucket/test/BucketListTests.cpp (hunk 5 above)
Comment (graydon) [reply src/bucket/test/BucketListTests.cpp:159]: fixed

File: src/bucket/test/BucketListTests.cpp (hunk 6 above)
Comment (graydon) [reply src/bucket/test/BucketListTests.cpp:646]: it's a helper that prints the value in two forms (decimal and hex).

File: src/util/Logging.cpp (hunk 7 above)
Comment (graydon) [reply src/util/Logging.cpp:301]: oh, yeah, oops. fixed.

**Comment by MonsieurNicolas** [comment]
r+ 6da163dafdde5868030c87aa0cf4bd4c6033b148

**Pull request merged**